### DISCUSSION TITLE
How work/manipulate rows data

### DISCUSSION THREAD

**Discussion by berusnet** [tdslite/tdslite#35] [Q&A] [answered]
Hello, recently i started to play with a s3 t-display but i'm noob in this field(CPP) and trying to learn/understand.
I have used tls library to insert and select data in a sql server.
I would like to see an example on how  work with result data.
Right now for example my  callback function is this 
```cpp
static void row_callback(void * u, const tdsl::tds_colmetadata_token & colmd,
                         const tdsl::tdsl_row & row) {

TFT_PRINTF_PROGMEM("%.*s %d %.*s",row [0].as<tdsl::char_view>().size(),row [0].as<tdsl::char_view>().data(),
                                    row [1].as<tdsl::int32_t>(),row[2].as<tdsl::char_view>().size(),
                                    row [2].as<tdsl::char_view>().data());
tdsl::char_view posta = row [0].as<tdsl::char_view>();
tdsl::int32_t qt = row [1].as<tdsl::int32_t>();
String pst{posta.data(), posta.size_bytes()};
SERIAL_PRINTLNF("TEST %s",pst);
SERIAL_PRINTLNF("TESTONE %d",qt); 
}
```

So i have 3 column. Which is the best way to assign row[0],row[1],row[2] to 3 arrays or array of object?
I would like to check if row[1][1]>row[1][2] o if it's equal to .. 2 or sum them ..make something before output.

Thanks in advance

**Comment by xmkg** [comment]
Hello @berusnet, thank you for the question.

Query results are directly delivered to the callback function "as-is", meaning that the data in the result network packet is not copied or altered in any way, and that is the what callback function sees when a result arrives[0]. Therefore, the text columns, such as TEXT, varchar, char, etc. are not NUL-terminated (i.e. they do not have `\0` at the end) so they're not "strings" if we're talking in C/C++ terms. They're an array of characters or character[1] spans if you will.  So, if you treat them as strings (e.g. print them with the `%s` format specifier, or try to copy them with `strcpy`) the program will not behave as expected.

So, back to your question. You need to create 3 string objects from the char_views:

```c++

#include <String.h>

static void row_callback(void * user_ptr, const tdsl::tds_colmetadata_token & colmd,
                         const tdsl::tdsl_row & row) {
    // re-cast user_ptr to its original type
    String (&strings) [3] = *reinterpret_cast<String(*)[3]>(user_ptr);
    
    int str_index = 0;
    for(tdsl::size_t i = 0; i < row.size(); i++){
        if(colmd.columns[i].type == tdsl::detail::e_tds_data_type::BIGVARCHRTYPE){
            tdsl::char_view vw = row[i].as<tdsl::char_view>();
            strings[str_index++] = String(vw.data(), vw.size());
        }
    }
}

void loop(){
    String strings[3];
    // pass `strings` as user data to the callback function
    driver.execute_query(tdsl::string_view{"SELECT a,b,c FROM test_rpc"}, row_callback, &strings);
}

```


[0]: This is done in order to eliminate unnecessary copying of data to reduce the memory footprint of the library.
[1]: The size of the "character" is not fixed (i.e. not always "1") and it depends on the column data type. For example, `VARCHAR` uses fixed-size characters[2] (i.e. "1") whereas `NVARCHAR` uses dynamic-sized characters, where a character size can be anything like 1,2,4, etc.
[2]: `VARCHAR` itself is represented as `BIGVARCHAR` at the binary protocol level.

**Reply by berusnet** [reply]
Thank you for your reply and patience.
As first test i have copy and paste your code but got 
![immagine](https://github.com/mustafakemalgilor/tdslite/assets/9430304/102ed7d5-8232-495c-83da-bb1b1a370fa5)

However i will try to clarify my question.
The callback function that i posted work perfectly and the output is right formatted.
In this two line of code
<code>tdsl::char_view posta = row [0].as<tdsl::char_view>();
tdsl::int32_t qt = row [1].as<tdsl::int32_t>();</code>
We get all the data of column 0 in 'posta' and column 1 in qt . In your code also strings[0] contain al the data of column 0
But if i want refer to check if at row 2 column 1 i have a value >100. or if i want check if at row 3 colunn 0 is equal to "ON" ?
For this reason i would like to convert 'qt' or 'posta' in to an array or if there are other solutions let me know.
Another question : You wrote "VARCHAR itself is represented as BIGVARCHAR at the binary protocol level." and INT is represented with ?

**Reply by xmkg** [reply, marked as answer]
The code is probably crashing because the fields are empty. If that's the case, the code below should fix it: 

```c++
static void row_callback(void * user_ptr, const tdsl::tds_colmetadata_token & colmd,
                         const tdsl::tdsl_row & row) {
    // re-cast user_ptr to its original type
    String (&strings) [3] = *reinterpret_cast<String(*)[3]>(user_ptr);
    
    int str_index = 0;
    for(tdsl::size_t i = 0; i < row.size(); i++){
        if(colmd.columns[i].type == tdsl::detail::e_tds_data_type::BIGVARCHRTYPE && row[i].size() > 0){
            tdsl::char_view vw = row[i].as<tdsl::char_view>();
            strings[str_index++] = String(vw.data(), vw.size());
        }
    }
}
```

> We get all the data of column 0 in 'posta' and column 1 in qt . In your code also strings[0] contain al the data of column 0
But if i want refer to check if at row 2 column 1 i have a value >100. or if i want check if at row 3 colunn 0 is equal to "ON" ?
For this reason i would like to convert 'qt' or 'posta' in to an array or if there are other solutions let me know.

You can do something similar to this:

```c++
tdsl::char_view posta = row [0].as<tdsl::char_view>();
tdsl::int32_t qt = row [1].as<tdsl::int32_t>();
tdsl::char_view col_ok = row [2].as<tdsl::char_view>();
tdsl::string_view OK {"OK"};
if(col_ok.size() == OK.size() && (memcmp(col_ok.data(), OK.data(), col_ok.size()) == 0) && qt > 100){
    // do some stuff...
}
```

> Another question : You wrote "VARCHAR itself is represented as BIGVARCHAR at the binary protocol level." and INT is represented with ?

It's [INTNTYPE](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/ce3183a6-9d89-47e8-a02f-de5a1a1303de).

**Reply by berusnet** [reply]
About the crash no field problem my mistake . forgot to comment an old line that execute another query.
So at the end the code work .
(edited : back to play with this after some days - was a dumb question) thanks of all
